### PR TITLE
cli: add internal tables to expose status server info about the cluster

### LINKED ISSUE #20713: cli/user,zone,node: run `user`, `zone`, `node` commands through SQL
Most `cli` commands individually query certain endpoints and compile a result which they then pretty-print (see, for example, [status node](https://github.com/cockroachdb/cockroach/blob/d1550657cb767a952e25d617c684a74bb3ebad8a/pkg/cli/node.go#L129)`.

This is suboptimal because it implies that certain kind of information is only exposed through the `cli`, and that extra development effort has to be undertaken to make the same status available through SQL when that is requested. Operators are usually able to connect to a cluster using SQL anyway, and using the `cli` for the same task may be cumbersome.

My proposal is to move the implementation of the `user`, `zone`, and `node` commands to SQL wholesale (so that the `cli` commands are just small wrappers that issue a SQL query).

Some of their information is already available in SQL and the rest should be, and in particular should be presented in a comparable way. For example, the various possible outputs of `./cockroach node status` should be simple joins of more elementary tables in `crdb_internal`.

Some care has to be taken to make sure that we support returning partial information (https://github.com/cockroachdb/cockroach/issues/16489). That is, it'd be unfortunate if we joined `crdb_internal.pretty_much_always_available_info_from_gossip` with `crdb_internal.this_only_works_when_cluster_healthy`. But these joins are small, and with just a hint of tooling the `cli` can do these in-memory while leaving blanks. Or, we make these tables itself return partial information, though that raises questions about timeouts, etc.

### PR DESCRIPTION
**Pull request by dsaveliev**
Fixes #20713.

Release note: None

### CODE REVIEW

File: pkg/sql/crdb_internal.go [hunk 1]
```diff
@@ -1613,8 +1621,8 @@ CREATE TABLE crdb_internal.gossip_nodes (
 	},
 }
 
-// crdbInternalGossipLiveness exposes local information about the nodes liveness.
-var crdbInternalGossipLiveness = virtualSchemaTable{
```
Comment (tschottdorf) [inline pkg/sql/crdb_internal.go]: nodes' liveness

**Review by tbg — COMMENTED** [review]
Looks good! Thanks for doing this. LGTM after minor nits.

File: pkg/sql/logictest/testdata/logic_test/crdb_internal [hunk 2]
```diff
@@ -228,17 +228,28 @@ node_id  component  field   value
...
-node_id  network  address           attrs  locality  version
-1        tcp      127.0.0.1:<port>  []     ·         <version>
+node_id  network  address           attrs  locality  server_version
+1        tcp      127.0.0.1:<port>  []     {}        <server_version>
 
 query IITBB colnames
 SELECT node_id, epoch, regexp_replace(expiration, '^\d+\.\d+,\d+$', '<timestamp>') as expiration, draining, decommissioning FROM crdb_internal.gossip_liveness WHERE node_id = 1
 ----
 node_id  epoch  expiration    draining  decommissioning
 1        1      <timestamp>   false     false
```
Comment (tschottdorf) [inline pkg/sql/logictest/testdata/logic_test/crdb_internal]: Could you add queries here and below that retrieve things like node_id, store_id, attrs, locality? Those should be static and so we'll see at least some data.

File: pkg/sql/crdb_internal.go [hunk 3]
```diff
@@ -1733,3 +1741,223 @@ CREATE TABLE crdb_internal.partitions (
...
+  go_version     STRING NOT NULL,
+  tag            STRING NOT NULL,
+  time           STRING NOT NULL,
+  revision       STRING NOT NULL,
+  cgo_compiler   STRING NOT NULL,
+  platform       STRING NOT NULL,
+  distribution   STRING NOT NULL,
+  type           STRING NOT NULL,
+  dependencies   STRING NOT NULL,
+  started_at     INT NOT NULL,
```
Comment (tschottdorf) [inline pkg/sql/crdb_internal.go]: TIMESTAMP

File: pkg/sql/crdb_internal.go [hunk 4]
```diff
@@ -1733,3 +1741,223 @@ CREATE TABLE crdb_internal.partitions (
...
+				metrics := json.NewBuilder()
+				for k, v := range s.Metrics {
+					metric, err := json.FromFloat64(v)
+					if err != nil {
+						return err
+					}
+					metrics.Add(k, metric)
+				}
+
+				percentilesToJSON := func(ps roachpb.Percentiles) (json.JSON, error) {
```
Comment (tschottdorf) [inline pkg/sql/crdb_internal.go]: @justinj do you think this code is good as is?

Comment (justinj) [reply pkg/sql/crdb_internal.go]: Assuming it's reasonable that these columns are JSON in the first place (based on a quick glance it looks like it is) the code constructing them looks good to me!

File: pkg/sql/crdb_internal.go (hunk 1 above)
Comment (dsaveliev) [reply pkg/sql/crdb_internal.go]: Shame on me. Fixed.

File: pkg/sql/logictest/testdata/logic_test/crdb_internal (hunk 2 above)
Comment (dsaveliev) [reply pkg/sql/logictest/testdata/logic_test/crdb_internal]: Sure, implemented.

**Comment by tschottdorf** [comment]
:lgtm: after my comments!

---

Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/crdb_internal.go, line 1745 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21427#-L2wTS6S8xSBlAmU-7HJ:-L2wTS6S8xSBlAmU-7HK:bp9f7gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a650593548851761e5c470a41d815b06f7d2be/pkg/sql/crdb_internal.go#L1745)):*
> ```Go
> }
> 
> // crdbInternalKVNodeStatusTable exposes an information from the status server about the cluster nodes.
> ```

`exposes information`

---

*[pkg/sql/crdb_internal.go, line 1858 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21427#-L2wTQTd6JELnz4qsx1Y:-L2wTQTe0XpeqQWPXtP7:bp9f7gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a650593548851761e5c470a41d815b06f7d2be/pkg/sql/crdb_internal.go#L1858)):*
> ```Go
> }
> 
> // crdbInternalKVStoreStatusTable exposes an information about the cluster stores.
> ```

`exposes information`

---

*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 243 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21427#-L2wT43c779wvpE9U803:-L2wT43c779wvpE9U804:bcrmosv) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a650593548851761e5c470a41d815b06f7d2be/pkg/sql/logictest/testdata/logic_test/crdb_internal#L243)):*
> ```
> 
> query ITTTTTTTTTTT colnames
> SELECT node_id, network, regexp_replace(address, '\d+$', '<port>') as address, attrs, locality, regexp_replace(server_version, '^\d+\.\d+(-\d+)?$', '<server_version>') as server_version, regexp_replace(go_version, '^go.+$', '<go_version>') as go_version, tag, time, revision, type, dependencies
> ```

`tag`, `time`, `revision`, `type`, `dependencies` will all be different. Just remove them :smile: 

---

*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 254 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21427#-L2wTGcy1bKQw2Rfp58Z:-L2wTGcy1bKQw2Rfp58_:b-5qb3yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a650593548851761e5c470a41d815b06f7d2be/pkg/sql/logictest/testdata/logic_test/crdb_internal#L254)):*
> ```
> ----
> node_id  store_id  attrs  capacity   available  used  range_count  writes_per_second
> 1        1         []     536870912  536870912  0     1            0
> ```

Remove `capacity` and `available` and `used`; they'll be different in each run. Also remove `range_count` and `writes_per_second`; these may differ.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21427#-:-L2wTWMimUyZ6Tvi25cB:b8pzn0)*

**Comment by a-robinson** [comment]
Great stuff, @dsaveliev!

**Pull request merged**